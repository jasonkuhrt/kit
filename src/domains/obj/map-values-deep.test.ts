import { Obj } from '#obj'
import { Test } from '#test'

// Test visitors
const errorToMessage = (value: any) => {
  if (value instanceof Error) return value.message
}

const truncateLongStrings = (value: any) => {
  if (typeof value === 'string' && value.length > 10) {
    return value.slice(0, 10) + '...'
  }
}

const doubleNumbers = (value: any) => {
  if (typeof value === 'number') return value * 2
}

const replaceSpecificObject = (value: any) => {
  if (typeof value === 'object' && value !== null && 'replace' in value) {
    return 'REPLACED'
  }
}

// dprint-ignore
Test.on(Obj.mapValuesDeep)
  .describeInputs('primitives pass through', [
    ['hello',      () => undefined],
    [42,           () => undefined],
    [true,         () => undefined],
    [null,         () => undefined],
    [undefined,    () => undefined],
  ])
  .describeInputs('early exit on match', [
    [new Error('test error'),                errorToMessage],
    ['short',                                truncateLongStrings],
    ['this is a very long string',           truncateLongStrings],
    [5,                                      doubleNumbers],
    [{ num: 10, nested: { num: 20 } },       doubleNumbers],
  ])
  .describeInputs('array recursion', [
    [[1, 2, 3],                              doubleNumbers],
    [[new Error('e1'), new Error('e2')],     errorToMessage],
    [['short', 'this is long'],              truncateLongStrings],
  ])
  .describeInputs('object recursion', [
    [{ a: 5, b: { c: 10 } },                 doubleNumbers],
    [{ err: new Error('test') },             errorToMessage],
    [{ str: 'short and long string' },       truncateLongStrings],
  ])
  .describeInputs('mixed structures', [
    [{ nums: [1, 2], nested: { nums: [3, 4] } },  doubleNumbers],
  ])
  .describeInputs('early exit stops recursion', [
    [{ replace: true, nested: { untouched: 'data' } },  replaceSpecificObject],
  ])
  .test()
