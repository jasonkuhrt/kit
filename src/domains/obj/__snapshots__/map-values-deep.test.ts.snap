// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`array recursion > mapValuesDeep([ 'short', 'this is long' ], (value) => { if (typeof value === "string" && value.length > 10) { return value....) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[ 'short', 'this is long' ]
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "string" && value.length > 10) {
    return value.slice(0, 10) + "...";
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS ARRAY
[ 'short', 'this is long' ]
╚══════════════════════════════════════════════════╝"
`;

exports[`array recursion > mapValuesDeep([ [Error: e1], [Error: e2] ], (value) => { if (value instanceof Error) return value.message; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[ [Error: e1], [Error: e2] ]
────────────────────────────────────────────────────
(value) => {
  if (value instanceof Error) return value.message;
}
╠══════════════════════════════════════════════════╣ THEN RETURNS ARRAY
[ 'e1', 'e2' ]
╚══════════════════════════════════════════════════╝"
`;

exports[`array recursion > mapValuesDeep([ 1, 2, 3 ], (value) => { if (typeof value === "number") return value * 2; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
[ 1, 2, 3 ]
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "number") return value * 2;
}
╠══════════════════════════════════════════════════╣ THEN RETURNS ARRAY
[ 1, 2, 3 ]
╚══════════════════════════════════════════════════╝"
`;

exports[`early exit on match > mapValuesDeep('short', (value) => { if (typeof value === "string" && value.length > 10) { return value....) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
short
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "string" && value.length > 10) {
    return value.slice(0, 10) + "...";
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
short
╚══════════════════════════════════════════════════╝"
`;

exports[`early exit on match > mapValuesDeep('this is a very long string', (value) => { if (typeof value === "string" && value.length > 10) { return value....) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
this is a very long string
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "string" && value.length > 10) {
    return value.slice(0, 10) + "...";
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
this is a very long string
╚══════════════════════════════════════════════════╝"
`;

exports[`early exit on match > mapValuesDeep([Error: test error], (value) => { if (value instanceof Error) return value.message; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
Error: test error
────────────────────────────────────────────────────
(value) => {
  if (value instanceof Error) return value.message;
}
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
test error
╚══════════════════════════════════════════════════╝"
`;

exports[`early exit on match > mapValuesDeep({ num: 10, nested: { num: 20 } }, (value) => { if (typeof value === "number") return value * 2; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  num: 10,
  nested: {
    num: 20
  }
}
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "number") return value * 2;
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  num: 10,
  nested: {
    num: 20
  }
}
╚══════════════════════════════════════════════════╝"
`;

exports[`early exit on match > mapValuesDeep(5, (value) => { if (typeof value === "number") return value * 2; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
5
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "number") return value * 2;
}
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
5
╚══════════════════════════════════════════════════╝"
`;

exports[`early exit stops recursion > mapValuesDeep({ replace: true, nested: { untouched: 'data' } }, (value) => { if (typeof value === "object" && value !== null && "replace" in val...) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  replace: true,
  nested: {
    untouched: 'data'
  }
}
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "object" && value !== null && "replace" in value) {
    return "REPLACED";
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
REPLACED
╚══════════════════════════════════════════════════╝"
`;

exports[`mixed structures > mapValuesDeep({ nums: [ 1, 2 ], nested: { nums: [ 3, 4 ] } }, (value) => { if (typeof value === "number") return value * 2; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  nums: [ 1, 2 ],
  nested: {
    nums: [ 3, 4 ]
  }
}
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "number") return value * 2;
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  nums: [ 1, 2 ],
  nested: {
    nums: [ 3, 4 ]
  }
}
╚══════════════════════════════════════════════════╝"
`;

exports[`object recursion > mapValuesDeep({ a: 5, b: { c: 10 } }, (value) => { if (typeof value === "number") return value * 2; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  a: 5,
  b: {
    c: 10
  }
}
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "number") return value * 2;
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  a: 5,
  b: {
    c: 10
  }
}
╚══════════════════════════════════════════════════╝"
`;

exports[`object recursion > mapValuesDeep({ err: [Error: test] }, (value) => { if (value instanceof Error) return value.message; }) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  err: [Error: test]
}
────────────────────────────────────────────────────
(value) => {
  if (value instanceof Error) return value.message;
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  err: 'test'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`object recursion > mapValuesDeep({ str: 'short and long string' }, (value) => { if (typeof value === "string" && value.length > 10) { return value....) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
{
  str: 'short and long string'
}
────────────────────────────────────────────────────
(value) => {
  if (typeof value === "string" && value.length > 10) {
    return value.slice(0, 10) + "...";
  }
}
╠══════════════════════════════════════════════════╣ THEN RETURNS OBJECT
{
  str: 'short and long string'
}
╚══════════════════════════════════════════════════╝"
`;

exports[`primitives pass through > mapValuesDeep('hello', () => void 0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
hello
────────────────────────────────────────────────────
() => void 0
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
hello
╚══════════════════════════════════════════════════╝"
`;

exports[`primitives pass through > mapValuesDeep(42, () => void 0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
42
────────────────────────────────────────────────────
() => void 0
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
42
╚══════════════════════════════════════════════════╝"
`;

exports[`primitives pass through > mapValuesDeep(null, () => void 0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
────────────────────────────────────────────────────
() => void 0
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`primitives pass through > mapValuesDeep(true, () => void 0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
true
────────────────────────────────────────────────────
() => void 0
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`primitives pass through > mapValuesDeep(undefined, () => void 0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
undefined
────────────────────────────────────────────────────
() => void 0
╠══════════════════════════════════════════════════╣ THEN RETURNS UNDEFINED
undefined
╚══════════════════════════════════════════════════╝"
`;
